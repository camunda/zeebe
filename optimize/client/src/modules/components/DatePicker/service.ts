/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {isValid, parseISO} from 'date-fns';
import {format} from 'dates';

export const DATE_FORMAT = 'yyyy-MM-dd';

export function isDateValid(date: string) {
  const parsedDate = parseISO(date);
  return isValid(parsedDate) && format(parsedDate, DATE_FORMAT) === date;
}
