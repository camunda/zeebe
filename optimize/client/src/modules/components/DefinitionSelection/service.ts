/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {Tenant} from 'types';
import {get, post} from 'request';

export type Version = {
  version: string;
  versionTag: string | null;
};

type DefintionWithTenants = {
  key: string;
  versions: string[];
  tenants: Tenant[];
};

type DefintionWithVersions = {key: string; versions: string[]};

export async function loadVersions(
  type: string,
  collectionId: string | null,
  key: string
): Promise<Version[]> {
  const params: {filterByCollectionScope?: string} = {};
  if (collectionId) {
    params.filterByCollectionScope = collectionId;
  }

  const response = await get(`api/definition/${type}/${key}/versions`, params);

  return response.json();
}

export async function loadTenants(
  type: string,
  definitions: DefintionWithVersions[],
  collectionId?: string | null
): Promise<DefintionWithTenants[]> {
  const payload: {filterByCollectionScope?: string; definitions: DefintionWithVersions[]} = {
    definitions,
  };
  if (collectionId) {
    payload.filterByCollectionScope = collectionId;
  }

  const response = await post(`api/definition/${type}/_resolveTenantsForVersions`, payload);

  return response.json();
}
