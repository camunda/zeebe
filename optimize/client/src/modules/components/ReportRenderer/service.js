/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';

export function isEmpty(str) {
  return !str || 0 === str.length;
}

export function getFormatter(measure) {
  if (typeof measure === 'object') {
    // can only happen for variable reports
    return formatters.frequency;
  }

  switch (measure) {
    case 'frequency':
      return formatters.frequency;
    case 'duration':
      return formatters.duration;
    case 'percentage':
      return formatters.percentage;
    default:
      return (v) => v;
  }
}

export const formatValue = (value, measure, precision) => {
  return getFormatter(measure)(value, precision);
};
