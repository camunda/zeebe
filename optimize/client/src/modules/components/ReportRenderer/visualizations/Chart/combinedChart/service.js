/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';

const {formatReportResult} = formatters;

// Override the default generate legend's labels function
// This is done to modify the colors retrieval method of the legend squares and filter unneeded labels
export function generateLegendLabels(chart) {
  const data = chart.data;
  return data.datasets.length
    ? data.datasets
        .map(function (dataset) {
          return {
            text: dataset.label,
            fillStyle: !dataset.backgroundColor.length
              ? dataset.backgroundColor
              : dataset.legendColor,
            strokeStyle: dataset.legendColor,
          };
        }, this)
        .filter((dataset) => {
          return dataset.text;
        })
    : [];
}

export function getCombinedChartProps(reports, data, measureIdx) {
  return data.reports.reduce(
    (prev, {id, color}) => {
      const report = reports[id];
      // skip unauthorized reports
      if (!report) {
        return prev;
      }
      let singleReportResult;
      if (data.visualization === 'number') {
        singleReportResult = [{key: report.id, value: report.result.data}];
      } else {
        singleReportResult = formatReportResult(data, report.result.measures[measureIdx].data);
      }

      return {
        resultArr: [...prev.resultArr, singleReportResult],
        reportsNames: [...prev.reportsNames, report.name],
        reportColors: [...prev.reportColors, color],
      };
    },
    {resultArr: [], reportsNames: [], reportColors: []}
  );
}
