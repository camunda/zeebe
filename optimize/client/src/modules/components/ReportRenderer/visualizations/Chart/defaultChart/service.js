/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';

const {convertToMilliseconds} = formatters;

export function getFormattedTargetValue({unit, value}) {
  if (!unit) {
    return value;
  }
  return convertToMilliseconds(value, unit);
}
