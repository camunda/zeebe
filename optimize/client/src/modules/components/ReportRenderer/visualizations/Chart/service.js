/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {getTooltipText, isDurationReport} from 'services';
import {t} from 'translation';

export function formatTooltip({
  dataset,
  dataIndex,
  configuration: {alwaysShowAbsolute, alwaysShowRelative, precision},
  formatter,
  instanceCount,
  isDuration,
  showLabel,
}) {
  if (dataset.isTarget) {
    return;
  }

  const shortNotation = isDuration ? alwaysShowAbsolute : alwaysShowAbsolute || alwaysShowRelative;

  const tooltipText = getTooltipText(
    dataset.data[dataIndex],
    formatter,
    instanceCount,
    alwaysShowAbsolute,
    alwaysShowRelative,
    isDuration,
    precision,
    shortNotation
  );

  if (!tooltipText) {
    return;
  }

  const label = showLabel && dataset.label ? dataset.label + ': ' : '';

  return shortNotation ? tooltipText : label + tooltipText;
}

export function formatTooltipTitle(title, availableWidth) {
  if (!title || !availableWidth) {
    return '';
  }

  const widthPerCharacter = 7;
  const charactersPerLine = Math.floor(availableWidth / widthPerCharacter);

  const lines = [];
  let remainingString = title;

  while (remainingString.length > charactersPerLine) {
    const currentString = remainingString.substr(0, charactersPerLine);
    let lastSpaceIndex = currentString.lastIndexOf(' ');
    const hasSuitableWhitespace = lastSpaceIndex !== -1;

    if (!hasSuitableWhitespace) {
      lastSpaceIndex = charactersPerLine;
    }

    lines.push(currentString.substr(0, lastSpaceIndex));
    remainingString = remainingString.substr(lastSpaceIndex + hasSuitableWhitespace);
  }
  lines.push(remainingString);

  return lines.join('\n');
}

export function getTooltipLabelColor({dataIndex, dataset}, type) {
  if (type === 'pie') {
    const color = dataset.backgroundColor[dataIndex];
    return {
      borderColor: color,
      backgroundColor: color,
    };
  }

  return {
    borderColor: dataset.legendColor,
    backgroundColor: dataset.legendColor,
  };
}

export function calculateLinePosition(chart) {
  const firstMeasureAxis = chart.scales['axis-0'];

  return firstMeasureAxis.getPixelForValue(chart.options.lineAt);
}

export function canBeInterpolated({type, value}, xml, decisionDefinitionKey) {
  if (type === 'flowNodes' || type === 'userTasks') {
    return false;
  }
  if (type === 'variable' && value.type === 'String') {
    return false;
  }
  if (type === 'inputVariable' || type === 'outputVariable') {
    return (
      new DOMParser()
        .parseFromString(xml, 'text/xml')
        .querySelector(
          `decision[id="${decisionDefinitionKey}"] [id="${value.id}"] ${
            type === 'inputVariable' ? 'inputExpression' : ''
          }`
        )
        .getAttribute('typeRef')
        .toLowerCase() !== 'string'
    );
  }
  return true;
}

export function getLabel({property, aggregationType, userTaskDurationTime}) {
  return (
    (userTaskDurationTime
      ? `${t('report.config.userTaskDuration.' + userTaskDurationTime)} `
      : '') +
    t('report.view.' + (property === 'frequency' ? 'count' : 'duration')) +
    (aggregationType
      ? ` - ${t('report.config.aggregationShort.' + aggregationType.type, {
          value: aggregationType.value,
        })}`
      : '')
  );
}

export function getAxisIdx(measures, measureIdx) {
  if (measures.every(({property}) => property === measures[0].property)) {
    // if every measure has the same prop, there is only one axis
    return 0;
  }
  return measures[measureIdx].property === 'frequency' ? 0 : 1;
}

export function hasReportPersistedTooltips(report) {
  const {
    data: {
      configuration: {alwaysShowAbsolute, alwaysShowRelative},
    },
  } = report;
  const isDuration = isDurationReport(report);
  return isDuration ? alwaysShowAbsolute : alwaysShowAbsolute || alwaysShowRelative;
}

export function getAxesConfig(isHorizontal) {
  return {
    axis0: {
      position: isHorizontal ? 'bottom' : 'left',
      axis: isHorizontal ? 'x' : 'y',
      id: 'axis-0',
    },
    axis1: {position: isHorizontal ? 'top' : 'right', axis: isHorizontal ? 'x' : 'y', id: 'axis-1'},
    groupBy: {
      position: isHorizontal ? 'left' : 'bottom',
      axis: isHorizontal ? 'y' : 'x',
      id: isHorizontal ? 'y' : 'x',
    },
  };
}
