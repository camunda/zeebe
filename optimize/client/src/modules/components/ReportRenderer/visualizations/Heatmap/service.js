/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {formatters} from 'services';
const {convertToMilliseconds} = formatters;

export function calculateTargetValueHeat(durationData, targetValues) {
  const data = {};

  Object.keys(targetValues).forEach((element) => {
    const targetValueInMs = convertToMilliseconds(
      targetValues[element].value,
      targetValues[element].unit
    );

    if (durationData[element] > targetValueInMs) {
      data[element] = durationData[element] / targetValueInMs - 1;
    } else {
      data[element] = null;
    }
  });

  return data;
}

export function getConfig(data, flowNodeId) {
  const {
    definitions,
    configuration: {heatmapTargetValue},
  } = data;
  const {value, unit} = heatmapTargetValue.values[flowNodeId];

  return {
    processDefinitionKey: definitions[0].key,
    processDefinitionVersions: definitions[0].versions,
    tenantIds: definitions[0].tenantIds,
    includedColumns: ['processInstanceId'],
    filter: [
      ...data.filter,
      {
        type: 'flowNodeDuration',
        data: {
          [flowNodeId]: {
            operator: '>',
            value,
            unit,
          },
        },
        filterLevel: 'instance',
      },
    ],
  };
}
