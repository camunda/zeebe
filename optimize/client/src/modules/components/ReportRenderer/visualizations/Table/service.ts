/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {post} from 'request';

export async function loadObjectValues(
  name: string,
  processInstanceId: string,
  processDefinitionKey: string,
  processDefinitionVersions: string[],
  tenantIds: (string | null)[]
): Promise<string> {
  const response = await post(`api/variables/values`, {
    name,
    processInstanceId,
    processDefinitionKey,
    processDefinitionVersions,
    tenantIds,
    type: 'object',
  });

  const values = await response.json();
  return values[0];
}
