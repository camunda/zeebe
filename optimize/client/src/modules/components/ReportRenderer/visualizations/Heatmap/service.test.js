/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {calculateTargetValueHeat, getConfig} from './service';

jest.mock('heatmap.js', () => {});
jest.mock('services', () => ({
  formatters: {convertToMilliseconds: (value) => value},
}));

describe('calculateTargetValueHeat', () => {
  it('should return the relative difference between actual and target value', () => {
    expect(calculateTargetValueHeat({a: 10}, {a: {value: 5, unit: 'millis'}})).toEqual({a: 1});
  });

  it('should return null for an element that is below target value', () => {
    expect(calculateTargetValueHeat({a: 2}, {a: {value: 5, unit: 'millis'}})).toEqual({a: null});
  });
});

it('should construct rawdata report with the target value as a filter', () => {
  const configuration = {
    heatmapTargetValue: {
      values: {
        flowNodeA: {
          value: 1234,
          unit: 'days',
        },
      },
    },
  };

  expect(
    getConfig(
      {
        configuration,
        definitions: [
          {
            key: '1',
            versions: ['1'],
            tenantIds: ['tenantA'],
          },
        ],
        filter: [{type: 'test'}],
      },
      'flowNodeA'
    )
  ).toMatchSnapshot();
});
