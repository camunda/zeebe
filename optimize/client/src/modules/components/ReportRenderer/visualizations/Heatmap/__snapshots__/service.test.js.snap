// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should construct rawdata report with the target value as a filter 1`] = `
Object {
  "filter": Array [
    Object {
      "type": "test",
    },
    Object {
      "data": Object {
        "flowNodeA": Object {
          "operator": ">",
          "unit": "days",
          "value": 1234,
        },
      },
      "filterLevel": "instance",
      "type": "flowNodeDuration",
    },
  ],
  "includedColumns": Array [
    "processInstanceId",
  ],
  "processDefinitionKey": "1",
  "processDefinitionVersions": Array [
    "1",
  ],
  "tenantIds": Array [
    "tenantA",
  ],
}
`;
